.github/workflows/functional.yml

```diff
@@ -0,0 +1,136 @@
+name: Functional Testing
+
+on:
+  push:
+    branches:
+      - main
+      - release-*
+    tags:
+      - "v[0-9]+.[0-9]+.[0-9]+*"
+  pull_request:
+
+
+defaults:
+  run:
+    shell: bash
+
+permissions:
+  contents: read
+
+jobs:
+  functional-tests:
+    name: Gateway Functional Tests
+    permissions:
+      contents: write # for docker/build-push-action to read repo content
```

Comment:
```suggestion
```

Do we need `write` permissions here? The comment mentions `read` only, and looking at the rest of the workflow, it doesn't look like we do need write so I don't believe this step is necessary (we have a top level read permissions already).